cumulusci/tasks/bulkdata/upsert_utils.py

```diff
@@ -0,0 +1,127 @@
+import typing as T
+
+from simple_salesforce import Salesforce
+from sqlalchemy import Column, MetaData, Table, Unicode, UniqueConstraint, and_, select
+from sqlalchemy.engine.base import Connection
+from sqlalchemy.exc import IntegrityError
+
+from cumulusci.core.exceptions import BulkDataException
+from cumulusci.tasks.bulkdata.mapping_parser import MappingStep
+from cumulusci.tasks.bulkdata.utils import sf_query_to_table
+
+
+def select_for_upsert(
+    *,
+    mapping: MappingStep,
+    metadata: MetaData,
+    connection: Connection,
+    context,
+):
+    """Pull data out of Salesforce for upsert matching
+
+    1.  Pull the data out into a temp table
+    2.  Make a query that joins the temp table to the
+        user-provided table by the update_key
+    """
+    upsert_key_table = _create_upsert_key_table(
+        mapping.sf_object, mapping.update_key, context, metadata, connection
+    )
+    upsert_query = _create_upsert_join_query(mapping, upsert_key_table, metadata)
+    return upsert_query
+
+
+def _create_upsert_join_query(
+    mapping: MappingStep,
+    upsert_key_table: Table,
+    metadata: MetaData,
+):
+    """Make a query that joins user-provided data to Salesforce IDs"""
+    update_data_table = metadata.tables[mapping.table]
+
+    relevant_data_columns = (
+        [update_data_table.c.id.label("_cci_local_id")]
+        + [update_data_table.c[key] for key in mapping.fields]
+        + [upsert_key_table.c.Id]
+    )
+
+    join_on_clauses = [
+        (
+            upsert_key_table.columns[column_name]
+            == update_data_table.columns[column_name]
+        )
+        for column_name in mapping.update_key
+    ]
+
+    q = select(*relevant_data_columns)
+
+    q = q.outerjoin(upsert_key_table, and_(*join_on_clauses))
+    return q
+
+
+def _create_empty_upsert_key_table(
+    tablename, metadata, fieldnames: T.List[str]
+) -> Table:
+    """Create an empty table matching the upsert semantics"""
+    if tablename not in metadata.tables:
+        fields = [Column(fieldname, Unicode(255)) for fieldname in fieldnames]
+        non_id_fields = [
+            fieldname for fieldname in fieldnames if fieldname.lower() != "id"
+        ]
+        assert len(fields) == len(non_id_fields) + 1
+        t = Table(tablename, metadata, *fields, UniqueConstraint(*non_id_fields))
+        t.create(metadata.bind)
+        return t
+
+
+def _create_upsert_key_table(
```

Comment:
This function name could more clearly state that we are actually performing a (potentially very lengthy) extraction. `_extract_upsert_key_data()`, perhaps?